File: matlab/tests/test_weight_profiles.m
%% test_weight_profiles.m — weight profiles beyond the window family
%
%  weightEvents accepts three kinds of profile: a numeric gamma (the
%  fixed-variance convolution family, tested in test_maet.m), one of
%  seven named profiles, and any function handle of the centred
%  difference. The named profiles absorb the serial-position curves
%  that seqWeights supplied before v3, so the values pinned here are
%  the ones that function produced: a rate of 1 over event numbers
%  1, ..., N decays by a factor of e per event.
%
%  Mirror of Python tests/test_weight_profiles.py; the two pin the same
%  numbers.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    addpath(fileparts(fileparts(mfilename('fullpath'))));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

wpIdx   = [1 2 3 4];
wpPitch = [60 62 64 65];
wpTol   = 1e-12;

wpPm      = preMaet({wpPitch, wpIdx});
wpRecency = exp(-[3 2 1 0]);

%% --- The anchored serial-position profiles ---------------------------

results{end+1, 1} = 'weight profiles: exponentialFromEnd is the old recency curve';
results{end, 2}   = ...
    max(abs(localProfileWeights(wpPm, NaN, 'exponentialFromEnd') - wpRecency)) < wpTol;

results{end+1, 1} = 'weight profiles: exponentialFromStart mirrors it';
results{end, 2}   = ...
    max(abs(localProfileWeights(wpPm, NaN, 'exponentialFromStart') ...
            - fliplr(wpRecency))) < wpTol;

results{end+1, 1} = 'weight profiles: decayRate and sd are two spellings of one scale';
results{end, 2}   = ...
    max(abs(localProfileWeights(wpPm, NaN, 'exponentialFromEnd', 'decayRate', 0.5) ...
                            - localProfileWeights(wpPm, NaN, ...
                                  'exponentialFromEnd', 'sd', 2))) < wpTol;

wpU = localProfileWeights(wpPm, NaN, 'uShape');
results{end+1, 1} = 'weight profiles: uShape is symmetric, and alpha selects its components';
results{end, 2}   = max(abs(wpU - fliplr(wpU))) < wpTol ...
    && max(abs(localProfileWeights(wpPm, NaN, 'uShape', 'alpha', 1) ...
               - localProfileWeights(wpPm, NaN, 'exponentialFromStart'))) < wpTol ...
    && max(abs(localProfileWeights(wpPm, NaN, 'uShape', 'alpha', 0) ...
               - localProfileWeights(wpPm, NaN, 'exponentialFromEnd'))) < wpTol;

wpExpectedAsym = 0.4 * exp(-0.2 * (wpIdx - wpIdx(1))) ...
    + 0.6 * exp(-0.8 * (wpIdx(end) - wpIdx));
results{end+1, 1} = 'weight profiles: uAsym gives its two components their own rates';
results{end, 2}   = ...
    max(abs(localProfileWeights(wpPm, NaN, 'uAsym', 'decayRateStart', 0.2, ...
                                'decayRateEnd', 0.8, 'alpha', 0.4) ...
                            - wpExpectedAsym)) < wpTol;

results{end+1, 1} = 'weight profiles: uAsym rates fall back to decayRate';
results{end, 2}   = max(abs(localProfileWeights(wpPm, NaN, 'uAsym', 'decayRate', 0.5) ...
                            - localProfileWeights(wpPm, NaN, 'uAsym', ...
                                  'decayRateStart', 0.5, ...
                                  'decayRateEnd', 0.5))) < wpTol;

%% --- The centre-anchored profiles ------------------------------------

results{end+1, 1} = 'weight profiles: exponentialBefore at the last event matches the anchored form';
results{end, 2}   = ...
    max(abs(localProfileWeights(wpPm, 4, 'exponentialBefore') - wpRecency)) < wpTol;

wpAfter = localProfileWeights(wpPm, 2, 'exponentialAfter');
results{end+1, 1} = 'weight profiles: exponentialAfter is zero below its centre';
results{end, 2}   = wpAfter(1) == 0 ...
    && max(abs(wpAfter(2:end) - exp(-[0 1 2]))) < wpTol;

results{end+1, 1} = 'weight profiles: the two-sided exponential holds its standard deviation';
results{end, 2}   = max(abs(localProfileWeights(wpPm, 2, 'exponential', 'sd', 1) ...
                            - exp(-abs(wpIdx - 2) * sqrt(2)))) < wpTol;

%% --- Function handles -------------------------------------------------

results{end+1, 1} = 'weight profiles: a function handle supplies any profile';
results{end, 2}   = max(abs(localProfileWeights(wpPm, 4, @(d) exp(0.5 * d) .* (d <= 0)) ...
                            - exp(0.5 * [-3 -2 -1 0]))) < wpTol;

results{end+1, 1} = 'weight profiles: a function handle is exempt from the kernel truncation';
results{end, 2}   = ...
    max(abs(localProfileWeights(wpPm, 1, @(d) 1 + abs(d)) - (1 + abs(wpIdx - 1)))) < wpTol;

%% --- Rejected combinations -------------------------------------------

results{end+1, 1} = 'weight profiles: a centre is refused by the anchored profiles';
results{end, 2}   = ...
    throwsErrorWithId(@() localProfileWeights(wpPm, 4, 'exponentialFromEnd'), ...
    'weightEvents:centreWithAnchoredProfile');

results{end+1, 1} = 'weight profiles: width is refused by the named profiles';
results{end, 2}   = ...
    throwsErrorWithId(@() localProfileWeights(wpPm, NaN, ...
        'exponentialFromEnd', 'width', 2), ...
    'weightEvents:widthWithNamedProfile');

results{end+1, 1} = 'weight profiles: sd and decayRate together are refused';
results{end, 2}   = throwsErrorWithId( ...
    @() localProfileWeights(wpPm, NaN, 'exponentialFromEnd', 'sd', 1, 'decayRate', 1), ...
    'weightEvents:sdRateXor');

results{end+1, 1} = 'weight profiles: asymmetric rates are refused by the one-rate profiles';
results{end, 2}   = throwsErrorWithId( ...
    @() localProfileWeights(wpPm, NaN, 'uShape', 'decayRateStart', 0.5), ...
    'weightEvents:asymRatesWithSymmetricProfile');

results{end+1, 1} = 'weight profiles: a rate is refused by the numeric family';
results{end, 2}   = ...
    throwsErrorWithId(@() localProfileWeights(wpPm, 2, 0, 'sd', 1, 'decayRate', 1), ...
    'weightEvents:rateWithNumericShape');

results{end+1, 1} = 'weight profiles: a scale is refused with a function handle';
results{end, 2}   = throwsErrorWithId( ...
    @() localProfileWeights(wpPm, 2, @(d) ones(size(d)), 'sd', 1), ...
    'weightEvents:scaleWithProfile');

results{end+1, 1} = 'weight profiles: an unknown profile name is refused';
results{end, 2}   = ...
    throwsErrorWithId(@() localProfileWeights(wpPm, NaN, 'exponentialFromMiddle'), ...
    'weightEvents:badShapeName');

results{end+1, 1} = 'weight profiles: a handle must return one non-negative factor per event';
results{end, 2}   = throwsErrorWithId(@() localProfileWeights(wpPm, 2, @(d) ones(1, 2)), ...
    'weightEvents:profileSize') ...
    && throwsErrorWithId(@() localProfileWeights(wpPm, 2, @(d) -ones(size(d))), ...
    'weightEvents:profileNotNonNegative');

%% --- Composition ------------------------------------------------------

wpPm = weightEvents(preMaet({wpPitch, wpIdx}), 2, 1, NaN, ...
                    'exponentialFromEnd', 'dropInputAttr', true);
[wpPOut, wpWOut, wpSpecsOut] = unpackPreMaet(wpPm);
results{end+1, 1} = 'weight profiles: the driving attribute is dropped and the pitches survive';
results{end, 2}   = numel(wpPOut) == 1 && numel(wpSpecsOut) == 1 ...
    && isequal(wpPOut{1}, wpPitch) ...
    && max(abs(wpWOut{1} - wpRecency)) < wpTol;

wpSalience = [1 0.6 0.6 1.3];
wpPmS = weightEvents(preMaet({wpPitch, wpIdx}, {wpSalience, []}), 2, 1, NaN, ...
                     'exponentialFromEnd', 'dropInputAttr', true);
[~, wpWS, ~] = unpackPreMaet(wpPmS);
results{end+1, 1} = 'weight profiles: an existing weight is multiplied into, not replaced';
results{end, 2}   = max(abs(wpWS{1} - wpSalience .* wpRecency)) < wpTol;

if standalone
    nPass = 0; nFail = 0;
    for i = 1:size(results, 1)
        if results{i,2}
            nPass = nPass + 1;
            fprintf('  PASS  %s\n', results{i,1});
        else
            nFail = nFail + 1;
            fprintf('  FAIL  %s\n', results{i,1});
        end
    end
    fprintf('\n=== test_weight_profiles: %d passed, %d failed (of %d) ===\n\n', ...
        nPass, nFail, nPass + nFail);
    clear cleanupDefaults
    if nFail > 0
        error('test_weight_profiles:failed', '%d test(s) failed.', nFail);
    end
end

function w = localProfileWeights(pm0, centre, shape, varargin)
%LOCALPROFILEWEIGHTS  Weight profile applied to pitch via an event number.
%   Drives a weightEvents profile on the two-attribute pre-MAET PM0
%   (pitch, event number) from the event-number attribute, drops that
%   attribute, and returns the pitch weights.
    pm = weightEvents(pm0, 2, 1, centre, shape, ...
                      'dropInputAttr', true, varargin{:});
    [~, wOut, ~] = unpackPreMaet(pm);
    w = wOut{1};
end

File: python/tests/test_weight_profiles.py
"""Weight profiles beyond the rectangle-Gaussian window family.

``weight_events`` accepts three kinds of profile: a numeric ``gamma``
(the fixed-variance convolution family, tested in ``test_maet.py``),
one of seven named profiles, and any callable of the centred
difference. The named profiles absorb the serial-position curves that
``seq_weights`` supplied before v3, so the values pinned here are the
ones that function produced: a rate of 1 over event numbers
1, ..., N decays by a factor of e per event.

The MATLAB twin is ``matlab/tests/test_weight_profiles.m``; the two
pin the same numbers.
"""
import numpy as np
import pytest

import mpt


IDX = np.array([[1.0, 2.0, 3.0, 4.0]])
PITCH = np.array([[60.0, 62.0, 64.0, 65.0]])


def _weights(shape, centre=None, **kw):
    """Profile applied to pitch through an event-number attribute."""
    pm = mpt.pack_pre_maet([PITCH, IDX])
    out = mpt.weight_events(pm, 1, 0, centre, shape,
                            drop_input_attr=True, **kw)
    return np.asarray(out["w_attr"][0]).ravel()


# --- the anchored serial-position profiles ---------------------------

def test_exponential_from_end_is_the_old_recency_curve():
    got = _weights("exponentialFromEnd")
    np.testing.assert_allclose(got, np.exp(-np.array([3.0, 2.0, 1.0, 0.0])))


def test_exponential_from_start_mirrors_it():
    np.testing.assert_allclose(_weights("exponentialFromStart"),
                               _weights("exponentialFromEnd")[::-1])


def test_decay_rate_and_sd_are_two_spellings_of_one_scale():
    np.testing.assert_allclose(_weights("exponentialFromEnd", decay_rate=0.5),
                               _weights("exponentialFromEnd", sd=2.0))


def test_u_shape_is_symmetric_and_alpha_selects_its_components():
    u = _weights("uShape")
    np.testing.assert_allclose(u, u[::-1])
    np.testing.assert_allclose(_weights("uShape", alpha=1.0),
                               _weights("exponentialFromStart"))
    np.testing.assert_allclose(_weights("uShape", alpha=0.0),
                               _weights("exponentialFromEnd"))


def test_u_asym_gives_its_two_components_their_own_rates():
    got = _weights("uAsym", decay_rate_start=0.2, decay_rate_end=0.8,
                   alpha=0.4)
    idx = IDX.ravel()
    expected = (0.4 * np.exp(-0.2 * (idx - idx[0]))
                + 0.6 * np.exp(-0.8 * (idx[-1] - idx)))
    np.testing.assert_allclose(got, expected)


def test_u_asym_rates_fall_back_to_decay_rate():
    np.testing.assert_allclose(_weights("uAsym", decay_rate=0.5),
                               _weights("uAsym", decay_rate_start=0.5,
                                        decay_rate_end=0.5))


# --- the centre-anchored profiles ------------------------------------

def test_exponential_before_matches_the_anchored_form_at_the_last_event():
    np.testing.assert_allclose(
        _weights("exponentialBefore", centre=4.0),
        _weights("exponentialFromEnd"))


def test_exponential_after_is_zero_below_its_centre():
    got = _weights("exponentialAfter", centre=2.0)
    assert got[0] == 0.0
    np.testing.assert_allclose(got[1:], np.exp(-np.array([0.0, 1.0, 2.0])))


def test_two_sided_exponential_holds_its_standard_deviation():
    # The Laplace standard deviation is tau * sqrt(2), so sd fixes the
    # variance as it does across the convolution family.
    got = _weights("exponential", centre=2.0, sd=1.0)
    d = IDX.ravel() - 2.0
    np.testing.assert_allclose(got, np.exp(-np.abs(d) * np.sqrt(2.0)))


# --- callables --------------------------------------------------------

def test_a_callable_supplies_any_profile():
    got = _weights(lambda d: np.exp(0.5 * d) * (d <= 0), centre=4.0)
    np.testing.assert_allclose(got, np.exp(0.5 * np.array([-3.0, -2.0, -1.0, 0.0])))


def test_a_callable_is_exempt_from_the_kernel_truncation():
    # A profile that grows with distance would be zeroed by the window
    # truncation; it survives, because an arbitrary profile need not decay.
    got = _weights(lambda d: 1.0 + np.abs(d), centre=1.0)
    np.testing.assert_allclose(got, 1.0 + np.abs(IDX.ravel() - 1.0))


# --- rejected combinations -------------------------------------------

def test_a_centre_is_refused_by_the_anchored_profiles():
    with pytest.raises(ValueError, match="anchors itself"):
        _weights("exponentialFromEnd", centre=4.0)


def test_width_is_refused_by_the_named_profiles():
    with pytest.raises(TypeError, match="support of the rectangle"):
        _weights("exponentialFromEnd", width=2.0)


def test_sd_and_decay_rate_together_are_refused():
    with pytest.raises(TypeError, match="two spellings"):
        _weights("exponentialFromEnd", sd=1.0, decay_rate=1.0)


def test_asymmetric_rates_are_refused_by_the_one_rate_profiles():
    with pytest.raises(TypeError, match="uAsym"):
        _weights("uShape", decay_rate_start=0.5)


def test_a_rate_is_refused_by_the_numeric_family():
    with pytest.raises(TypeError, match="rectangle-Gaussian"):
        _weights(0.0, centre=2.0, decay_rate=1.0)


def test_a_scale_is_refused_with_a_callable():
    with pytest.raises(TypeError, match="carries its own scale"):
        _weights(lambda d: np.ones_like(d), centre=2.0, sd=1.0)


def test_an_unknown_profile_name_is_refused():
    with pytest.raises(ValueError, match="Unknown profile"):
        _weights("exponentialFromMiddle")


def test_a_callable_must_return_one_non_negative_factor_per_event():
    with pytest.raises(ValueError, match="one factor per event"):
        _weights(lambda d: np.ones(2), centre=2.0)
    with pytest.raises(ValueError, match="non-negative"):
        _weights(lambda d: -np.ones_like(d), centre=2.0)


# --- composition ------------------------------------------------------

def test_the_driving_attribute_is_dropped_and_the_pitches_survive():
    pm = mpt.pack_pre_maet([PITCH, IDX])
    out = mpt.weight_events(pm, 1, 0, None, "exponentialFromEnd",
                            drop_input_attr=True)
    p_out, w_out, specs = mpt.unpack_pre_maet(out)
    assert len(p_out) == 1 and len(specs) == 1
    np.testing.assert_allclose(np.asarray(p_out[0]), PITCH)
    np.testing.assert_allclose(np.asarray(w_out[0]).ravel(),
                               np.exp(-np.array([3.0, 2.0, 1.0, 0.0])))


def test_an_existing_weight_is_multiplied_into_not_replaced():
    salience = np.array([[1.0, 0.6, 0.6, 1.3]])
    pm = mpt.pack_pre_maet([PITCH, IDX], [salience, None])
    out = mpt.weight_events(pm, 1, 0, None, "exponentialFromEnd",
                            drop_input_attr=True)
    np.testing.assert_allclose(
        np.asarray(out["w_attr"][0]).ravel(),
        salience.ravel() * np.exp(-np.array([3.0, 2.0, 1.0, 0.0])))
